src/tls/verifier.rs

```diff
@@ -193,33 +192,25 @@ fn verify_presented_certs(presented_certs: &[Certificate]) -> Result<(), TLSErro
     certificate
         .check_self_issued()
         .map_err(TLSError::WebPKIError)?;
-    verify_libp2p_signature(&extension, certificate.subject_public_key_info().key())
+    verify_libp2p_signature(&extension, certificate.subject_public_key_info().spki())
         .map_err(TLSError::WebPKIError)
 }
 
-struct Libp2pExtension<'a> {
+struct Libp2pExtension {
     peer_key: PublicKey,
-    signature: &'a [u8],
+    signature: Vec<u8>,
 }
 
-fn parse_libp2p_extension(extension: Input<'_>) -> Result<Libp2pExtension<'_>, Error> {
-    fn read_bit_string<'a>(input: &mut Reader<'a>, e: Error) -> Result<Input<'a>, Error> {
-        der::bit_string_with_no_unused_bits(input).map_err(|_| e)
-    }
-
+fn parse_libp2p_extension(extension: Input<'_>) -> Result<Libp2pExtension, Error> {
     let e = Error::ExtensionValueInvalid;
-    Input::read_all(&extension, e, |input| {
-        der::nested(input, der::Tag::Sequence, e, |input| {
-            let public_key = read_bit_string(input, e)?.as_slice_less_safe();
-            let signature = read_bit_string(input, e)?.as_slice_less_safe();
-            // We deliberately discard the error information because this is
-            // either a broken peer or an attack.
-            let peer_key = PublicKey::from_protobuf_encoding(public_key).map_err(|_| e)?;
-            Ok(Libp2pExtension {
-                peer_key,
-                signature,
-            })
-        })
+    let (public_key, signature): (Vec<u8>, Vec<u8>) = yasna::decode_der(extension.as_slice_less_safe())
+        .map_err(|_| e)?;
```

Comment:
Please keep the original version unless there is a good reason to change it.  The code I wrote avoids extra copies and heap allocations, and I am far more confident in its robustness against malicious inputs.  In particular, it uses the same ASN.1 code as _ring_ and WebPKI.

Reply:
Here is what I get if I revert it:

```
Jul 30 15:15:26.878 DEBUG libp2p_quic::muxer: connection lost because of the cryptographic handshake failed: error 40: invalid certificate: ExtensionValueInvalid
```

So I tend to keep it.

Reply:
This is due to the Go implementation not conforming to the spec.  The specification [clearly states] that the extension is the following ASN.1 structure:

```asn1
SignedKey ::= SEQUENCE {
  publicKey BIT STRING,
  signature BIT STRING 
}
```

and that is what my code implements.  However, the Go code uses the following (incompatible) ASN.1 structure:

```asn1
SignedKey ::= SEQUENCE {
  publicKey OCTET STRING,
  signature OCTET STRING 
}
```

You can test this by replacing line 207 of the original code with

```rust
        // The specification states that this is a BIT STRING, but the Go implementation
        // uses an OCTET STRING.  OCTET STRING is superior in this context, so use it.
        der::expect_tag_and_get_value(input, der::Tag::OctetString).map_err(|_| e)
```

and reverting the lifetime-related changes.  This should interoperate successfully with the Go version, and I would be fine with it.

The `SignedKey` structure used by the Go implementation is strictly better than the one specified, as both `publicKey` and `signature` will _always_ be whole numbers of bytes.  Using `BIT STRING` wastes two bytes per `SignedKey` structure and is more error-prone.

[clearly states]: https://github.com/libp2p/specs/blob/98e0ca0e54fd83f7b8c86d7e9677e1e430da5810/tls/tls.md#libp2p-public-key-extension

Reply:
You were right. I reverted this code, applied your suggestion and it worked.

Reply:
Filed an issue: https://github.com/libp2p/specs/pull/357